src/sorted_container_iteration.jl

```diff
@@ -0,0 +1,1167 @@
+
+
+```julia
+   for (t,k) in tokens(keys(exclusive_key(sd, key1, key2)))
+      <body>
+   end
+```
+Here, `sd` is a `SortedDict` or `SortedMultiDict`, `key1` and `key2` are keys
+indexing `sd`.  In this case, `t` will be tokens of consecutive items,
+while `k` will be the corresponding keys.  The returned keys lie between `key1` and
+`key2` excluding `key2`.
+
+```julia
+   for (t,k) in Iterators.reverse(tokens(keys(exclusive_key(sd, key1, key2))))
+      <body>
+   end
+```
+Same as above, except the iteration is in the reverse order.
+
+Running time for all iterations: O(*c*(*s* + log *n*)), where
+*s* is the number of steps from start to end of the iteration.
+"""
+function Base.iterate(s::SortedContainerIterable,
+                      state = iteration_init(s))
+    if state.next == state.final
+        return nothing
+    else
+        return (get_item(s, state), next(s, state))
+    end
+end
+
+Base.keytype(::IterableObject{C, R, KeysValsIter, NoTokens, D}) where
+{C <: SDMContainer, R, D} = keytype(C)
+
+Base.keytype(::Type{IterableObject{C, R, KeysValsIter, NoTokens, D}}) where
+{C <: SDMContainer, R, D} = keytype(C)
+
+Base.valtype(::IterableObject{C, R, KeysValsIter, NoTokens, D}) where
+{C <: SDMContainer, R, D} = valtype(C)
+
+Base.valtype(::Type{IterableObject{C, R, KeysValsIter, NoTokens, D}}) where
+{C <: SDMContainer, R, D} = valtype(C)
+
+Base.IteratorSize(::Type{T} where {T <: SortedContainer}) = HasLength()
+Base.IteratorSize(::Type{IterableObject{C, EntireContainer, KV, T, D}}) where
+{C, KV, T, D} = HasLength()
+
+Base.IteratorSize(::Type{IterableObject{C, ExclusiveRange, KV, T, D}}) where
+{C, KV, T, D} = SizeUnknown()
+
+Base.IteratorSize(::Type{IterableObject{C, InclusiveRange, KV, T, D}}) where
+{C, KV, T, D} = SizeUnknown()
+
+Base.length(ito::IterableObject{C, EntireContainer, KV, T, D}) where
+{C, KV, T, D} = length(ito.m)
+
+
+"""
+    Base.in(x, iter::SortedContainerIterable)
+
+Returns true if `x` is in `iter`, where `iter` refers to any of the
+iterable objects described under [`Base.iterate(iter::SortedContainerIterable)`](@ref),
+and `x` is of the appropriate type. For all of the iterables
+except the five listed below, the algorithm used is a linear-time
+search. For example, the call:
+
+    (k=>v) in exclusive(sd, st1, st2)
+
+where `sd` is a SortedDict, `st1` and `st2` are semitokens, `k` is a
+key, and `v` is a value, will loop over all entries in the
+dictionary between the two tokens and a compare for equality using
+`isequal` between the indexed item and `k=>v`.
+
+The five exceptions are:
+
+```julia
+(k=>v) in sd
+(k=>v) in smd
+k in ss
+k in keys(sd)
+k in keys(smd)
+```
+
+Here, `sd` is a SortedDict, `smd` is a SortedMultiDict, and `ss` is
+a SortedSet.
+
+These five invocations of `in` use the index structure of the sorted
+container and test equality based on the order object of the keys
+rather than `isequal`. Therefore, these five are all faster than
```

Comment:
should we comment here that if `isequal` is defined consistently with `isless` etc this should have same behavour?

Reply:
The issue here is not whether `isless` and `isequal` are consistent.  Rather, the issue is whether the user has defined an ordering (such as case-insensitive for strings) for the container different from `isequal` and `isless`.